pkg/controller/compute/v1beta1/cluster.go

```diff
@@ -0,0 +1,287 @@
+	for _, pool := range existing.NodePools {
+		if pool != nil {
+			if pool.Name == gke.BootstrapNodePoolName {
+				boostrapNP := strings.Join([]string{meta.GetExternalName(mg), gke.BootstrapNodePoolName}, "/")
+				_, err = e.cluster.Projects.Locations.Clusters.NodePools.Delete(boostrapNP).Context(ctx).Do()
+				if err != nil {
+					// If we fail to delete bootstrap node pool we will be requeued implicitly.
+					return resource.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errDeleteBootstrapNodePool)
+				}
+				break
+			}
+		}
+	}
+
+	switch cr.Status.AtProvider.Status {
+	case v1beta1.ClusterStateRunning:
+		cr.Status.SetConditions(v1alpha1.Available())
+		resource.SetBindable(cr)
+	case v1beta1.ClusterStateProvisioning:
+		cr.Status.SetConditions(v1alpha1.Creating())
+	case v1beta1.ClusterStateUnspecified, v1beta1.ClusterStateDegraded, v1beta1.ClusterStateError, v1beta1.ClusterStateReconciling:
+		cr.Status.SetConditions(v1alpha1.Unavailable())
+	}
+
+	upToDate, _ := gke.IsUpToDate(&cr.Spec.ForProvider, *existing)
+
+	return resource.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  upToDate,
+		ConnectionDetails: connectionDetails(existing),
+	}, nil
+}
+
+func (e *clusterExternal) Create(ctx context.Context, mg resource.Managed) (resource.ExternalCreation, error) {
+	i, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return resource.ExternalCreation{}, errors.New(errNotCluster)
+	}
+
+	// Generate GKE cluster from resource spec.
+	cluster := gke.GenerateCluster(i.Spec.ForProvider)
+
+	// Insert default node pool for bootstrapping cluster.
+	gke.GenerateNodePoolForCreate(cluster)
+
+	create := &container.CreateClusterRequest{
+		Cluster: cluster,
+	}
+
+	parent := fmt.Sprintf(parentFormat, e.projectID, i.Spec.ForProvider.Location)
+
+	if _, err := e.cluster.Projects.Locations.Clusters.Create(parent, create).Context(ctx).Do(); err != nil {
+		return resource.ExternalCreation{}, errors.Wrap(err, errCreateCluster)
+	}
+
+	clusterName := fmt.Sprintf(clusterNameFormat, e.projectID, i.Spec.ForProvider.Location, i.Spec.ForProvider.Name)
+	meta.SetExternalName(i, clusterName)
+
+	// TODO(hasheddan): go ahead and propagate username / password here if set in spec?
+	return resource.ExternalCreation{}, nil
+}
+
+func (e *clusterExternal) Update(ctx context.Context, mg resource.Managed) (resource.ExternalUpdate, error) {
+	i, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return resource.ExternalUpdate{}, errors.New(errNotCluster)
+	}
+
+	// We have to get the cluster again here
+	existing, err := e.cluster.Projects.Locations.Clusters.Get(meta.GetExternalName(mg)).Context(ctx).Do()
+	if err != nil {
+		return resource.ExternalUpdate{}, errors.Wrap(err, errGetCluster)
+	}
+
+	upToDate, updateFn := gke.IsUpToDate(&i.Spec.ForProvider, *existing)
+	if upToDate {
+		return resource.ExternalUpdate{}, nil
+	}
+
+	_, err = updateFn(e.cluster, ctx, meta.GetExternalName(mg))
```

Comment:
Is there a way we could avoid leaking knowledge of the GKE service, network context, etc into our clients package? I prefer to keep all that contained here in the controller code (or more specifically, prefer to have that knowledge and responsibility in one package rather than split across two).

Perhaps `IsUpToDate` could return the required `&container.UpdateClusterRequest{}` rather than returning a function that makes the actual update call?

Reply:
Unfortunately the reason this is happening is that some fields require specific updates methods, so `IsUpToDate()` does not always return an update func that is the same or even uses `&container.UpdateClusterRequest`